client/network/src/service.rs

```diff
@@ -991,10 +991,17 @@ impl Metrics {
 	}
 }
 
-impl<B: BlockT + 'static, H: ExHashT> Future for NetworkWorker<B, H> {
-	type Output = Result<(), io::Error>;
+impl<B: BlockT + 'static, H: ExHashT> NetworkWorker<B, H> {
+	/// Performs one action on the network, then returns.
+	///
+	/// The returned future is designed to be freely cancellable.
+	pub async fn next_action(&mut self) {
+		future::poll_fn(move |cx| Pin::new(&mut *self).poll(cx)).await
+	}
 
-	fn poll(mut self: Pin<&mut Self>, cx: &mut std::task::Context) -> Poll<Self::Output> {
+	/// Implementation of `next_action`. Note that this is not an implementation of `Future` but
```

Comment:
As this is an implementation detail now, it could as well just implement future, no?

Reply:
I precisely don't want `NetworkWorker` to implement `Future`, otherwise it's impossible to rewrite that as an async block.